### DISCUSSION TITLE
Weekly Blog 6

### DISCUSSION THREAD

**Discussion by JAWolfs** [SE-TINF22B6/Underwatch#65] [Announcements]
Hello guys and girls,

This week, we want to share information on two themes:
1. the design principles we applied in our project and 
2. the progress we've made with our website.

# Design Principles
We want to present our Design-principles to you. Our architectural design is mainly based on the MVC (Model View Control) pattern. This extents to our game as-well as our backend. The main idea of the mvc-pattern is to seperate a program based on different roles. 
1. *The Model:* This is a collection of classes or structs, which model the data wich is used in the program.
2. *The View:* This is the part of the program which is responsible for the presentation of the program to the user. This could be a GUI, a website, or a terminal interface.
3. *The Controller:* This are the parts of the program which are responsible for the logic and the control flow of the program.

*Implementing this pattern has a few advantages for a program:*
1. *Clear Responsibility:* Every part of the program has a clear responsibility. This allows for better maintenance further down the road of the project. Because changes should optimally only affect certain parts of the program which will be easier to implement. Also adding new features should not need changing existing code responsible for other things. As an example, adding a new weapon to the game should not need changes for existing weapons.
2. *Modularity:* If we plan the program to be modular from the beginning, it allows for better abstractions of the code. Thus we can reuse many of the patterns for other parts of the code. This is especially visible in the architecture of the website, where we use and reuse many React-Compontents in this style.

# Website Progress
In the last two sprints (4 weeks), we focused on developing the initial version of our website. As you may have noticed, we created a very basic mock-up earlier. However, the current state of the website is much more advanced than the mock-up ever was. Currently, we have two main sections:

## The Landing Page
![Landing Page](https://github.com/SE-TINF22B6/Underwatch/assets/123720714/be533c03-e716-401c-a29d-ad0e2e2b96b5)

This serves as a simple introductory site for everyone. There isn't anything particularly special about it. It just displays "Underwatch" and highlights that it is the best game created by TINF22B6. Additionally, there are two buttons on the landing page. One directs you to the scoreboard, while the other currently serves no function. It is intended to lead you to a wiki that we plan to create. In the background, there's a blurred video playing in a loop, showcasing some early game footage of Underwatch. The color of the text and buttons is inspired by the light from the torches.

## The Scoreboard
![Scoreboard](https://github.com/SE-TINF22B6/Underwatch/assets/123720714/60458639-7e03-4715-8087-55cf7444573a)

As you can see, the background remains the same video as on the landing page. There's a table with currently three rows, and it's very likely that we'll expand that in the future. The background for the text and the table is in purple, a color taken from the floor tiles in the game. The text is the same color as before.

## The Scoreboard with Filters
![Filter](https://github.com/SE-TINF22B6/Underwatch/assets/123720714/d8856364-af54-45fd-bcc0-31d59856ff53)

Clicking the filter button on the scoreboard page opens a small box with options to filter the displayed results. Currently, the date is a text field, but we plan to change it to a date-picker in the future. However, nothing in the filter box is functional at the moment, and the scoreboard is also hardcoded.

**Comment by giomskii** [comment]
Good evening team underwatch,
i think it's very convenient to use the MVC-pattern for your project, since you can extract the elements of your game such as the player and environment models, the logic and the user interface. Are you also going to use other design patterns like the SOLID ones?
In regards to your initial website implementation, i think it looks good especially your choice of color and amount of ui elements/components. However, the title of the landing page tends to blend in with the video in the background. Nonetheless, nice progress and excited to see where this is heading 👍

**Comment by Shiny1708** [comment]
You described the Model View Controller pattern very well to give everyone reading a good understanding about the concept. I think using MVC is a very good approach for your project and suits a game very well. I have to agree with @FurKay00 on the landing page. Either the text or the background should have a different color for better readability.

**Comment by LeonmAndreesen** [comment]
Hi, 
i really think you really elaborated really well on what this model is. This helps even  others as a recap of the lecture. Moreover, telling the others the advantages of that model is a huge help. This leads to a clear understanding of what you are doing. 
I really like the layout of your pages. I can see why others already suggested some colorchanges. Would definetely help with seeing that website.

**Comment by AndresMasis** [comment]
Hi. First of all as asidenote, I really like your UI design. Before reading your post it already looks like a game I would like to play.

Now related to the content of your post.

 I agree that the MVC architecture is appropiate for this project. Since it is web hosted, I think it is a good approach.
I also think it was a good decision to focus in the modularity. I can imagine that you have character, world elements and so on. These should definitely be managed with a good modularity.

Cheers from CookPit.
